internal/envoy/versions.go

```diff
@@ -20,30 +20,122 @@ import (
 	"fmt"
 	"io"
 	"net/http"
+	"strconv"
+	"strings"
 
+	"github.com/tetratelabs/func-e/internal/globals"
 	"github.com/tetratelabs/func-e/internal/version"
 )
 
-// FuncEVersions returns a version map from a remote URL. eg globals.DefaultEnvoyVersionsURL.
-func FuncEVersions(ctx context.Context, envoyVersionsURL string, p version.Platform, v version.Version) (version.ReleaseVersions, error) {
+type funcEVersionsGetter func() (version.ReleaseVersions, error)
+
+// FuncEVersions fetches a version map form a remote URL and gets the latest patched version given
+// a version input. This input can be major.minor.patch (e.g. 1.1.1) or major.minor (1.1).
+type FuncEVersions interface {
+	// Get returns a version map from a remote URL. e.g. from globals.DefaultEnvoyVersionsURL.
+	Get() (version.ReleaseVersions, error)
+
+	// FindLatestVersions returns the latest patch version given an Envoy version.
+	FindLatestVersions(version.Version) (string, error)
+}
+
+// NewFuncEVersions creates a new FuncEVersions to fetch a version map from a remote URL.
+func NewFuncEVersions(ctx context.Context, envoyVersionsURL string, p version.Platform, v version.Version) FuncEVersions {
+	return &funcEVersions{ctx: ctx, envoyVersionsURL: envoyVersionsURL, platform: p, version: v}
+}
+
+type funcEVersions struct {
+	ctx              context.Context
+	envoyVersionsURL string
+	platform         version.Platform
+	version          version.Version
+
+	// Allow to override Get implementation.
+	getFunc funcEVersionsGetter
+}
+
+// Get returns a version map from a remote URL. e.g. from globals.DefaultEnvoyVersionsURL.
+func (f funcEVersions) Get() (version.ReleaseVersions, error) {
 	result := version.ReleaseVersions{}
 	// #nosec => This is by design, users can call out to wherever they like!
-	resp, err := httpGet(ctx, envoyVersionsURL, p, v)
+	resp, err := httpGet(f.ctx, f.envoyVersionsURL, f.platform, f.version)
 	if err != nil {
 		return result, err
 	}
 	defer resp.Body.Close() //nolint
 
 	if resp.StatusCode != http.StatusOK {
-		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, envoyVersionsURL)
+		return result, fmt.Errorf("received %v status code from %v", resp.StatusCode, f.envoyVersionsURL)
 	}
 	body, err := io.ReadAll(resp.Body) // fully read the response
 	if err != nil {
-		return result, fmt.Errorf("error reading %s: %w", envoyVersionsURL, err)
+		return result, fmt.Errorf("error reading %s: %w", f.envoyVersionsURL, err)
 	}
 
 	if err := json.Unmarshal(body, &result); err != nil {
 		return result, fmt.Errorf("error unmarshalling Envoy versions: %w", err)
 	}
+
 	return result, nil
 }
+
+// FindLatestVersions returns the latest patch version given an Envoy version.
+func (f funcEVersions) FindLatestVersions(envoyVersion version.Version) (string, error) {
```

Comment:
it makes me uncomfortable a little to pass in version.Version and pass out a string. I think we shouldn't enter this at all if we already know we have a concrete version. So, if anything it should pass out version.Version and in minorVersion

Also, we tend to use pointer style
```suggestion
// FindLatestPatch finds the latest patch version for the given minor or raises an error.
// minor is just like envoy.Version format, except missing the patch. Ex. 1.18 or 1.20_debug
func (f *funcEVersions) FindLatestPatch(minorVersion string) (version.Version, error) {
```